packages/jupyter-widgets/README.md

```diff
@@ -28,7 +28,27 @@ export default class MyNotebookApp extends ReactComponent {
 
 ## Documentation
 
-We're working on adding more documentation for this component. Stay tuned by watching this repository!
+The `jupyter-widgets` package supports two types of widgets:
+- Standard widgets provided in the official [`jupyter-widgets/base`](https://www.npmjs.com/package/@jupyter-widgets/base) and [`jupyter-widgets/controls`](https://www.npmjs.com/package/@jupyter-widgets/controls) package
+- [Custom Widgets](https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20Custom.html) or 3rd party widgets authored by the OSS community
+
+The `WidgetDisplay` component has an additional prop named `customWidgetLoader`  to provide custom loaders for fetching 3rd party widgets. A reference implementation for a custom loader which serves as the default for this package can be found in `widget-loader.ts`.
+
+```typescript
+customWidgetLoader?: (moduleName: string, moduleVersion: string) => Promise<any>;
+```
+
+### Custom Widgets
+
+Since custom widgets are hosted on CDN, we set https://unkpg.com as our default CDN Base URL. The default base URL can be overriden by specifying another URL via the HTML attribute "data-jupyter-widgets-cdn" on any script tag of the page.
+
+For instance if your js bundle is loaded as `bundle.js` on your page and you wanted to set [jsdelivr](https://www.jsdelivr.com) as your default CDN url for custom widgets, you could do the following:
+```html
+<script data-jupyter-widgets-cdn="https://cdn.jsdelivr.net/npm" src="bundle.js"></script>
```

Comment:
@rgbkrk Thoughts on this model for passing custom properties to the JavaScript assets?

Reply:
Interesting -- this makes it so a custom frontend can specify where custom widgets are allowed to be loaded from?

Probably worth noting that any output can add a tag to say to load other assets, so that's another vector for attack (though it means you need some HTML output already emitted in the document to use it).

Reply:
> Interesting -- this makes it so a custom frontend can specify where custom widgets are allowed to be loaded from?

Correct. This allows you to set the CDN to use when loading widgets. The HTML widget manager does something similar ([ref](https://github.com/jupyter-widgets/ipywidgets/blob/82a7959bd24cd520893f94dcd459aebbff7a4794/docs/source/embedding.md)).

> Probably worth noting that any output can add a tag to say to load other assets, so that's another vector for attack (though it means you need some HTML output already emitted in the document to use it).

Good point. The [logic](https://github.com/nteract/outputs/pull/47/files#diff-8018fc3c56df63e7c70bb905aa08c20beedc7b5be14305e22bc29e5cccf77225R73-R81) currently loads the the attribute from any script tag. I believe the way the for-loop is setup will cause it to favor the most recently inserted script tag which might be a problem.

@vivek1729 Thoughts?

Reply:
That's a good point @rgbkrk. @captainsafia rightly pointed out that I followed the pattern that the HTML widget manager to override the default CDN URL. The logic of reading and looping over `script` tags is also based off the HTML Widget manager ([ref](https://github.com/jupyter-widgets/ipywidgets/blob/master/packages/html-manager/src/libembed-amd.ts#L10)).

> Good point. The logic currently loads the the attribute from any script tag. I believe the way the for-loop is setup will cause it to favor the most recently inserted script tag which might be a problem.

Trying to understand how favoring the most recent added script tag might be a problem. Ideally, we'd hope that there's only 1 script tag that specifies the `data-jupyter-widgets-cdn` tag. Looping over script tags potentially opens up the possibility of script tags being added dynamically which can lead to the cdn URL being over-written. However, we should also note that the CDN url is set when the `WidgetManager` is getting constructed. Since the `WidgetManager` is a singleton, this would only happen once and it makes sense because we don't really want to have the base CDN URL change very often. It's more like a one-time setup configuration.

Reply:
The `WidgetManager` gets instantiated once when the first widget is rendered onto the page. It's possible for a nefarious to render a HTML output (or inject HTML into the page in some other way) that renders a script tag with a malicious CDN onto the page _then_ render a second output containing the ipywidget.

@jasongrout Do you have any thoughts on this since I believe you implemented the original code in the HTML manager?

Reply:
@captainsafia, thanks for the additional details. How do you suggest we address this concern?

Reply:
For now, I'm thinking that we:

- Merge this PR so we can get the basic groundwork in
- Ship an alpha release of the package
- Open an issue to address the security issue, particularly use the alpha package to create an MVP of what the attack would look like, also see if we can get some insight from the ipywidgets folks on this
- Followup on the issue from #3 as needed

If that sounds good to you we can move forward with it.

Reply:
@captainsafia , this sounds like a good plan to me.

Reply:
I'll proceed with the PR when you get a chance to do a final review and approve.